sw/otbn/crypto/sha3-shake.s

```diff
@@ -0,0 +1,753 @@
+/* Copyright lowRISC contributors (OpenTitan project).
+ * Copyright 2015 Markku-Juhani O. Saarinen.
+ * Use of this source code is governed by the MIT license 
+ * that can be found in the LICENSE.tiny_sha3 file.
+ */
+
+/**
+ * This library implements hash computation as specified in FIPS PUB 202
+ * "SHA-3 Standard: Permutation-based Hash and Exendable Output Function"
+ *
+ * The implementation is based on tiny_sha3 by Saarinen.
+ * https://github.com/mjosaarinen/tiny_sha3.
+ * Terminology within the comments in this library is based (as much as
+ * possible) on the terminology of tiny_sha3.
+ *
+ * SHA-3 context MUST be 32B-aligned. It includes:
+ *  -- 200B - data  - offset = 0
+ *  --   4B - pt    - offset = 200
+ *  --   4B - rsiz  - offset = 204
+ *  --   4B - mdlen - offset = 208
+ */
+
+.section .text
+/**
+ * sha3_init
+ *
+ * Initialize SHA-3 "context". 
+ *
+ * @param[in]  x11: Message digest length
+ * @param[out] x10: Pointer to 212 bytes of memory for holding the context.
+                   32B aligned.
+ *
+ * clobbered registers: x5 to x7
+ * clobbered flag groups: none
+ */
+.global sha3_init
+sha3_init:
+  addi x5, x10, 0
+  addi x6, x0, 31
+
+  /* Initialize the state to all zeros using w31:
+     -- 6 * 32  = 192 B 
+     -- 192 + 8 = 200 B: DONE */
+  LOOPI 6, 1
+    bn.sid x6, 0(x5++)
+  sw x0, 0(x5)
+  sw x0, 4(x5)
+
+  /* Save pt, mdlen, rsiz to memory */
+  sw   x0, 8(x5)
+  slli x6, x11, 1
+  addi x7, x0, 200
+  sub  x6, x7, x6
+  sw   x6, 12(x5)
+  sw   x11, 16(x5)
+  ret
+
+/**
+ * sha3_update
+ *
+ * Update state with additional data
+ *
+ * @param[in]    x12: Data length
+ * @param[in]    x11: Pointer to data
+ * @param[inout] x10: Pointer to 212 bytes of memory for holding the context.
+                   32B aligned.
+ *
+ * clobbered registers: x5 to x7, x13 to x17, x28 to x31, 
+ *                      w0 to w13, w21 to w30
+ * clobbered flag groups: FG0
+ */
+.global sha3_update
+sha3_update:
+  /* Load pt as j */
+  lw x5, 200(x10)
+  /* Load constant for address alignment */
+  li x30, 0xFFFFFFFC
+  /* Data counter */
+  li x13, 0
+
+  /* The data address is called the 'source address'; while the 'context address at pt is called the 'destination address'. We need to compute:
+  -- st[j++] ^= data[i] for i=0,..,x12
+  
+  We iterate over each input byte. As there is only 4B- or 32B-aligned load/store, unaligned memory access is tackled by masking the address with 0xFFFFFFFC to make it 4B-aligned. The the last two bits of the address is masked out and used to compute the position of the wanted byte. */
+  beq x12, x0, _sha3_update_skip_loop
+  LOOP x12, 22
+    /* Get destination address */
+    add  x6, x10, x5 
+    addi x5, x5, 1 
+    /* Align destination address and load data */
+    and  x7, x6, x30
+    lw   x28, 0(x7)
+    /* Mask out the lowest 2 bits of destination address
+       to compute position of st[j] */
+    andi x29, x6, 0x3
+    slli x29, x29, 3 
```

Comment:
Thanks for the comment! Yes, we considered that but the code gets quite a bit longer and more complicated because the alignment needs to be ensured upon starting to read/write (so we need to work our way from byte-wise -> 4B-wise -> 32B-wise) and then when finishing, we need to do the reverse to deal with remainders smaller than 32B or 4B. 
I'll give the implementation a try and see how much code it will be later this week.